packages/wasm/crate/src/planner.rs

```diff
@@ -1,591 +1,389 @@
+impl ActionList {
+    fn new() -> Self {
         Self {
-            rng,
-            balance: Balance::default(),
-            vote_intents: BTreeMap::default(),
-            plan: TransactionPlan::default(),
-            ibc_actions: Vec::new(),
-            gas_prices: GasPrices::zero(),
-            fee_tier: FeeTier::default(),
+            actions: Vec::new(),
+            change_outputs: BTreeMap::new(),
         }
     }
 
-    /// Set the current gas prices for fee prediction.
-    pub fn set_gas_prices(&mut self, gas_prices: GasPrices) -> &mut Self {
-        self.gas_prices = gas_prices;
-        self
-    }
-
-    /// Set the current fee tier.
-    pub fn set_fee_tier(&mut self, fee_tier: FeeTier) -> &mut Self {
-        self.fee_tier = fee_tier;
-        self
-    }
-
-    /// Get the current transaction balance of the planner.
-    pub fn balance(&self) -> &Balance {
-        &self.balance
+    fn balance(&self) -> Balance {
+        let mut balance = Balance::zero();
+        for action in &self.actions {
+            balance += action.balance();
+        }
+        for action in self.change_outputs.values() {
+            balance += action.balance();
+        }
+        balance
     }
 
-    /// Get all the note requests necessary to fulfill the current [`Balance`].
-    pub fn notes_requests(
-        &self,
-        source: Option<AddressIndex>,
-    ) -> (Vec<NotesRequest>, Vec<NotesForVotingRequest>) {
-        (
-            self.balance
-                .required()
-                .map(|Value { asset_id, amount }| NotesRequest {
-                    asset_id: Some(asset_id.into()),
-                    address_index: source.clone(),
-                    amount_to_spend: Some(amount.into()),
-                    include_spent: false,
-                })
-                .collect(),
-            self.vote_intents
-                .iter()
-                .map(
-                    |(
-                        _proposal, // The request only cares about the start block height
-                        VoteIntent {
-                            start_block_height, ..
-                        },
-                    )| NotesForVotingRequest {
-                        votable_at_height: *start_block_height,
-                        address_index: source.clone(),
-                    },
-                )
-                .collect(),
-        )
+    fn push(&mut self, action: ActionPlan) {
+        self.actions.push(action);
     }
 
-    /// Set the expiry height for the transaction plan.
-    pub fn expiry_height(&mut self, expiry_height: u64) -> &mut Self {
-        self.plan.transaction_parameters.expiry_height = expiry_height;
-        self
-    }
+    fn gas_estimate(&self) -> Gas {
+        // TODO: this won't include the gas cost for the bytes of the tx itself
+        // so this gas estimate will be an underestimate, but since the tx-bytes contribution
+        // to the fee is ideally small, hopefully it doesn't matter.
```

Comment:
cc @zbuc -- we could fix this in the protocol by attributing the block space costs to the action, and get a perfect estimate

Reply:
It doesn't look like we're even including any transaction-level byte cost right now FWIW:

```rust
impl GasCost for Transaction {
    fn gas_cost(&self) -> Gas {
        self.actions().map(GasCost::gas_cost).sum()
    }
}
```